src/core/src/pass/stateful_to_stateless.cpp

```diff
@@ -0,0 +1,172 @@
+    std::string variable_name;  // original variable_id
+    std::string input_name;     // restored name of input
+    std::string output_name;    // restored name of output
+};
+
+typedef std::vector<Variable> Variables;
+
+void restore_kv_cache_order(Variables& variables, const std::unordered_map<std::string, size_t>& var_index_by_var_id) {
+    // Try to restore variable order based on the known naming convention from optimum-intel.
+    // If names are not satisfy the expected convention, fallback to use order based on var_index_by_var_id
+    // Sort items that do satisfy the naming conventions before items that don't satisfy.
+
+    std::stable_sort(variables.begin(), variables.end(), [&](const Variable& a, const Variable& b) {
+        if (a.index >= 0 && b.index >= 0) {
+            return a.index < b.index;
+        } else if (a.index >= 0 && b.index < 0) {
+            return true;
+        } else if (a.index < 0 && b.index >= 0) {
+            return false;
+        } else {  // a.index < 0 && b.index < 0
+            return var_index_by_var_id.at(a.variable_name) < var_index_by_var_id.at(b.variable_name);
+        }
+    });
+}
+
+}  // namespace
+
+bool ov::pass::StatefulToStateless::run_on_model(const std::shared_ptr<ov::Model>& model) {
+    RUN_ON_MODEL_SCOPE(StatefulToStateless);
+
+    auto beam_idx = get_parameter_by_tensor_name(model, "beam_idx");
+    Variables variables;  // to collect variables corresponding to future_params
+    variables.reserve(model->get_sinks().size());
+    Variable::Context context;
+    std::unordered_map<std::string, std::shared_ptr<ov::Node>>
+        future_params;  // to collect nodes, each with a single output that will be replaced by new parameters
+    if (beam_idx) {
+        for (const ov::Input<ov::Node>& input : beam_idx->get_output_target_inputs(0)) {
+            if (auto gather = std::dynamic_pointer_cast<op::util::GatherBase>(input.get_node()->shared_from_this())) {
+                auto read_value =
+                    std::dynamic_pointer_cast<op::util::ReadValueBase>(gather->get_input_node_shared_ptr(0));
+                OPENVINO_ASSERT(read_value,
+                                "Unexpected model topology in StatefulToStateless: no ReadValue is found at the first "
+                                "input of Gather by `beam_idx` parameter");
+                auto variable_name = read_value->get_variable_id();
+                variables.push_back(Variable(context, variable_name));
+                future_params[variable_name] = gather;
+            }
+        }
+    } else {
+        OPENVINO_THROW(
+            "Stateful models without `beam_idx` input are not supported in StatefulToStateless transformation");
+    }
+    model->remove_parameter(beam_idx);
+
+    typedef std::shared_ptr<op::util::AssignBase> PAssign;
+    std::unordered_map<std::string, PAssign> assigns_by_var_id;
+    std::unordered_map<std::string, size_t> assign_index_by_var_id;
```

Comment:
As I understand it, we need this map only for "stable_sort" call, can we use assigns_by_var_id instead and call "get_variable_id" in the stable sort lambda?